### PR TITLE
Fix MultiConstraint optimization with version overlap

### PR DESCRIPTION
**Pull request by jderusse**
This is an alternative fix for https://github.com/composer/semver/commit/d07997aec5baf2463c7576b4d0e74d083f71f686#commitcomment-44111784

### CODE REVIEW

File: tests/Constraint/MultiConstraintTest.php [hunk 1]
```diff
@@ -339,92 +339,35 @@ public function multiConstraintOptimizations()
...
-                        ),
+                        new Constraint('>=', '1.0.0.0-dev'),
+                        new Constraint('<', '5.0.0.0-dev'),
+                        new Constraint('!=', '2.0.1.0'),
                     ),
-                    false
+                    true
                 )
             ),
             'Test does not collapse multiple contiguous with multiple other constraint' => array(
```
Comment (stof) [inline tests/Constraint/MultiConstraintTest.php:349]: the description of these test cases should probably be updated

Comment (jderusse) [reply tests/Constraint/MultiConstraintTest.php]: lazy me !

thanks

File: src/Constraint/MultiConstraint.php [hunk 2]
```diff
@@ -249,6 +247,29 @@ private static function optimizeConstraints(array $constraints, $conjunctive)
...
+                        $c = (string)$left->constraints[$j];
+                        switch ($c[0]) {
+                            case '!':
+                            case '>':
+                            case '=':
+                                break;
+                            default:
+                                continue 3;
+                        }
+                    }
```
Comment (Seldaek) [inline src/Constraint/MultiConstraint.php:271]: Not against doing this, but I think some comments would be good here :)

Why !/>/= and !/</= are OK and the rest not?

I see how != is fine to optimize as that's a single gap in the range. 

But I don't see why == is ok to optimize, `^1.0 || ^2.0, =2.0.3` means `^1.0 || 2.0.3` but with this optimization it'd become `1.0 - 3.0, 2.0.3` which is only matched by `2.0.3`. The `^1.0` is dropped.

I am also not sure why </> are ok in each respective loop. `^1.0 <1.1 || ^2.0 > 2.0.4` getting merged to `1.0 - 3.0, <1.1, > 2.0.4` also seems wrong to me as this suddenly cannot match anything anymore.

So all in all this seems broken, except for != which can be safely kept, but then the question is it worth looping over things just in the odd case we have a !=, which is anyway quite rare..

Comment (jderusse) [reply src/Constraint/MultiConstraint.php:271]: yeah.. you're right, I missed to much cases... let's close this

**Pull request closed without merging**